Apply some obvious rich debug info optimizations

jakobbotsch opened the pull request:
When rich debug info is enabled we send MethodDetails events for all inlinees as otherwise no information for those may have been sent (in case they have not been jitted). During rundown this was sending a lot of duplicate events, so use a hash set to avoid this. On Avalonia.ILSpy which has around ~33k methods jitted, we go from ~94k MethodDetails events to ~43k MethodDetails events on rundown.

In addition, compress rich debug info when stored in the runtime. For the above scenario the memory overhead of rich debug info goes from 7 MB -> 2.1 MB.

Also optimize the format we use for the ETW events. While we do not use the same delta compression as the runtime storage, we now avoid sending out padding and also use a smaller type for the offset mappings source type. The average size of a rich debug info ETW event goes from 259 bytes to 219 bytes for the above scenario.

The net result is the following for two rundowns of Avalonia.ILSpy (I clicked around a bit and did a few actions before enabling the rundown. EventID(189) is the rich debug info event)
Before:
![image](https://user-images.githubusercontent.com/7887810/182860458-a7232167-4f2b-4bf2-974a-3eb33812bcc5.png)

After:
![image](https://user-images.githubusercontent.com/7887810/182860632-4d7c9d16-f64e-4931-bb6d-ec20bfce579c.png)


There should not be any effect of this PR except when rich debug info is enabled.

I would like to get this in .NET 7. There is more work to be done here for .NET 8, e.g. by avoiding storing method handles when we can get it based on IL offset, but this fixes the most glaring issues.

jakobbotsch committed:
Stop sending many duplicate MethodDetails events for rich debug info

When rich debug info is enabled we send MethodDetails events for all
inlinees as otherwise no information for those may have been sent (in
case they have not been jitted). During rundown this was sending a lot
of duplicate events.

jakobbotsch committed:
Compress rich debug info and optimize ETW events

* Compress rich debug info when stored in the runtime. For
  Avalonia.ILSpy which has ~33k methods, the memory overhead of rich
  debug info goes from 7 MB -> 2.1 MB

* ETW events do not use the delta compression, but still optimize them a
  little by avoiding sending out the padding and unnecessarily large
  types. The average size of a rich debug info ETW event goes from 259
  bytes to 219 bytes for ILSpy

jakobbotsch committed:
Nit

jakobbotsch committed:
Another nit

jakobbotsch committed:
Fix x86 build

jakobbotsch committed:
Fix checked build

davidwrighton commented on src/coreclr/vm/eventtrace.cpp:
```diff
@@ -6976,20 +6993,48 @@ VOID ETW::MethodLog::SendMethodRichDebugInfo(MethodDesc* pMethodDesc, PCODE pNat
+        WriteToBuffer(&pBuffer, numMappings);
+        for (uint32_t i = 0; i < numInlineTree; i++)
+        {
+            WriteToBuffer(&pBuffer, inlineTree[i].Method);
```
Use of WriteToBuffer is a concern for me.

1. The definition of these types does not have comments/description that the field's types are part of a contract that is never permitted to change.
2. This is an inefficient encoding. For instance, ILOffset is unlikely to use the full 32 bits that is being sent here, and the Child/Sibling are highly unlikely to be large as well.

davidwrighton replied:
We discussed offline. I'm ok with the current set of changes.

jakobbotsch committed:
Add a contract

jakobbotsch committed:
Add DoEncodedDeltaU32NonMonotonic and use it

The inline ordinals may not be monotonically increasing and the existing
DoEncodedDeltaU32 is unnecessarily inefficient for the cases where it
isn't (plus, will assert).

jakobbotsch committed:
Use non-monotonic delta compression for IL offsets

Gets us an extra ~10% size reduction in my tests

jakobbotsch commented:
I added a `DoEncodedDeltaU32NonMonotonic` as I realized the existing `DoEncodedDeltaU32` assumes monotonicity, and I was hitting asserts when testing with checked builds since inlinee IDs may not be monotonic (this would translate to inefficient compression in release builds). I also made the IL offset encoding use it, which nets us another ~10% reduction in memory usage for the debug info.

I've verified that some basic debugging/stepping scenarios work, and that the diagnostics tests pass, with both rich debug info on and off.

jakobbotsch commented:
Failure is #73247
